Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp

```diff
@@ -653,86 +653,390 @@ bool TerrainSystem::GetIsHoleFromFloats(float x, float y, Sampler sampler) const
-                                                      AZ::Vector3(x + range, y, 0.0f),
-                                                      AZ::Vector3(x, y + range, 0.0f) };
+    // The full set of positions to query to be able to calculate all the normals.
+    AZStd::vector<AZ::Vector3> queryPositions;
+    queryPositions.reserve(inPositions.size() * queryCount);
 
-    AZStd::array<float, 4> heights = { 0.0f, 0.0f, 0.0f, 0.0f };
-    AZStd::array<bool, 4> exists = { false, false, false, false };
-    GetHeightsSynchronous(directionVectors, sampler, heights, exists);
+    // To determine if terrain exists at the queried normal, we'll use the terrain exists flag from the terrain grid point closest
+    // to the position. We'll track this while building up our query positions.
+    AZStd::vector<uint8_t> existsIndex;
+    existsIndex.reserve(inPositions.size());
 
-    directionVectors[0].SetZ(heights[0]);
-    directionVectors[1].SetZ(heights[1]);
-    directionVectors[2].SetZ(heights[2]);
-    directionVectors[3].SetZ(heights[3]);
+    for (const auto& position : inPositions)
+    {
+        // For each position, determine where in the square the point falls and get the bottom left corner of the square.
+        AZ::Vector2 normalizedDelta;
+        AZ::Vector2 bottomLeft;
+        ClampPosition(position.GetX(), position.GetY(), queryResolution, bottomLeft, normalizedDelta);
 
-    outNormal = (directionVectors[2] - directionVectors[1]).Cross(directionVectors[3] - directionVectors[0]).GetNormalized();
+        // Calculate the four corners of our grid square:
+        //  2 *-* 3
+        //    |\|
+        //  0 *-* 1
 
-    if (terrainExistsPtr)
+        AZStd::array<AZ::Vector3, 4> corners =
+        {
+            AZ::Vector3(bottomLeft.GetX()                  , bottomLeft.GetY(), 0.0f),
+            AZ::Vector3(bottomLeft.GetX() + queryResolution, bottomLeft.GetY(), 0.0f),
+            AZ::Vector3(bottomLeft.GetX()                  , bottomLeft.GetY() + queryResolution, 0.0f),
+            AZ::Vector3(bottomLeft.GetX() + queryResolution, bottomLeft.GetY() + queryResolution, 0.0f)
+        };
+
+        // Our grid squares are squares so the diagonal is a 45 degree angle. We can determine which triangle the query point
+        // falls in just by checking if (x + y) < 1. This arbitrarily assigns points on the diagonal to the upper triangle.
+        // We could use "<=" if we wanted them to go to the upper triangle.
+        bool bottomTriangle = (normalizedDelta.GetX() + normalizedDelta.GetY()) < 1.0f;
+
+        // We'll query for the 3 vertices of whichever triangle of the square the query point falls in.
+        // We'll order them in counter-clockwise order and follow a mirrored pattern between the two, so that we can calculate the
+        // normal from the results for either triangle by using (second - first) x (third - first).
+        if (bottomTriangle)
```

Comment:
nit - this probable could be simplified a bit by doing the "closest point" check independently of which triangle it is. If indices 0 and 1 were the shared corners of the two triangles, and index 2 meant the _not_ shared corner, you could check if it's closest to 0 or 1 first, otherwise it's index 2 for either top or bottom triangle.

Reply:
Discussed offline - simplified further by just setting "exists" to mean "all 3 points for the normal exist".